File: src/main/java/org/academy/api/client/gui/util/GlyphCommandGenerator.java
package org.academy.api.client.gui.util;

import org.academy.api.client.gui.command.DrawCommand;
import org.academy.api.client.gui.command.GlyphDrawCommand;
import org.academy.api.client.gui.msdf.core.MsdfConstants;
import org.academy.api.client.gui.msdf.layout.MsdfTextProcessor;

import java.util.ArrayList;
import java.util.List;

public final class GlyphCommandGenerator {
    private GlyphCommandGenerator() {
    }

    public static List<DrawCommand> generate(
            String text,
            float fontSize, float thickness,
            float red, float green, float blue, float alpha
    ) {
        List<DrawCommand> commands = new ArrayList<>();
        var instances = MsdfTextProcessor.layout(text, fontSize);

        for (var instance : instances) {

            commands.add(new GlyphDrawCommand(
                    instance.textureView(),
                    instance.x(),
                    instance.y(),
                    instance.quadWidth(),
                    instance.quadHeight(),
                    instance.u0(), instance.v0(), instance.u1(), instance.v1(),
                    red, green, blue, alpha,
                    MsdfConstants.DEFAULT_PX_RANGE,
                    thickness
            ));
        }

        return commands;
    }
}

File: src/main/kotlin/org/academy/api/client/gui/util/GlyphCommandGenerator.kt
package org.academy.api.client.gui.util

import org.academy.api.client.gui.command.GlyphDrawCommand
import org.academy.api.client.gui.msdf.Constants
import org.academy.api.client.gui.msdf.layout.MsdfTextProcessor.layout

object GlyphCommandGenerator {
    fun generate(
        text: String,
        fontSize: Float, thickness: Float,
        red: Float, green: Float, blue: Float, alpha: Float
    ): MutableList<GlyphDrawCommand> {
        val commands: MutableList<GlyphDrawCommand> = mutableListOf()
        val result = layout(text, fontSize)

        for (instance in result.instances) {
            commands.add(
                GlyphDrawCommand(
                    instance.textureView,
                    instance.x,
                    instance.y,
                    instance.quadWidth,
                    instance.quadHeight,
                    instance.u0, instance.v0, instance.u1, instance.v1,
                    red, green, blue, alpha,
                    Constants.DEFAULT_PX_RANGE,
                    thickness
                )
            )
        }

        return commands
    }
}
